### PR TITLE
[internals refactor] docs and reorganize compilation units

### PR DESCRIPTION
**Pull request by majestrate**
begin piecewise refactor

rearrange compilation units in cmake to have unambiguous component separation, with documentation notes on what the responsibilities of each are.

the new library structure:

* creates `lokinet-base` interface library that sets up common include paths and compile flags
* makes `lokinet-amalgum` an interface library with everything in it.
* marks the `lokinet-service-deprecated-kitchensink` component (the thing we are splitting up)
* defines the desired component layout going forward, adds stub libraries but no code or headers
* very brief documentation of component relations

to get the refactor smaller a few concessions have been made:

* have all our headers be referenced in a consistent albeit dumb way (consistency is preferred)

these commits in order are:

1) adding high level plan docs
2) redoing header conversion
3) cmake restructure
4) clang format comment

### CODE REVIEW

**Comment by majestrate** [comment]
blocked on review from @jagerman

File: CMakeLists.txt
Comment (jagerman) [inline CMakeLists.txt]: Why keep this?

File: docs/internals_overview.md
Comment (jagerman) [inline docs/internals_overview.md]: Not unclear: https://datatracker.ietf.org/doc/html/rfc9221

File: docs/internals_overview.md
Comment (jagerman) [inline docs/internals_overview.md]: Please ditch the hyperbole, keep this doc professional.

File: docs/refactor_notes.md
Comment (jagerman) [inline docs/refactor_notes.md]: Is there ever going to be more than one implementer of these interfaces?  I feel like we already have too many abstract interfaces that serve little purpose right now because they only have (and likely only ever will have) one single implementation.

Comment (jagerman) [reply docs/refactor_notes.md]: Additionally, for performance reasons, we ought to avoid unnecessary virtual inheritance, particularly for anything that gets used in hot code: there is a run-time cost (albeit small) of virtual lookup table access.

File: docs/refactor_notes.md
Comment (jagerman) [inline docs/refactor_notes.md]: FWIW I'm not a fan of the "I" prefix, partly because there really is no such thing as a pure interface in C++ (i.e. it's purely convention).

Proper interface classes (by convention in C++, or concrete in other languages) are awkwardly restricted in many classes.  For instance, I could easily imagine this `I_PlatformDetails` having some protected member variables/functions that de-duplicate common functionality from platform code, but as soon as you have any member or non-pure-virtual base implementation it isn't really an interface anymore.

(And many of our current `I_` classes *do* violate this principle, because having base implementations or members is both common and useful).

File: docs/refactor_notes.md
Comment (jagerman) [inline docs/refactor_notes.md]: Why not?

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt]: I feel like this belongs in `external/`

File: docs/internals_overview.md
Comment (jagerman) [inline docs/internals_overview.md]: Refactored how?

File: docs/internals_overview.md
Comment (jagerman) [inline docs/internals_overview.md]: This section feels needlessly verbose.  Would it not be sufficient to say:
"Lokinet uses a peer-based architecture, not a client-server architecture.  Thus we refer to interactions solely in terms of initiators and recipients, not clients and servers."
and *just* that.  All the other stuff feels irrelevant to the rest of this doc (e.g. alice and bob are never mentioned again), and I don't understand the need to bother justifying so heavily not using client-server terminology.  It feels like you're trying to make an argument against someone who strongly wants to use client-server terminology but such a person doesn't exist?

File: CMakeLists.txt
Comment (majestrate) [reply CMakeLists.txt]: so i break less things in CI. (lazy)

File: docs/internals_overview.md
Comment (majestrate) [reply docs/internals_overview.md]: gotcha

File: docs/internals_overview.md
Comment (majestrate) [reply docs/internals_overview.md]: yup, this was cherrypicked from a local branch

File: docs/refactor_notes.md
Comment (majestrate) [reply docs/refactor_notes.md]: fair

File: docs/refactor_notes.md
Comment (majestrate) [reply docs/refactor_notes.md]: so a `WhateverThing_Base` as an abstract class would be better then?

File: docs/refactor_notes.md
Comment (majestrate) [reply docs/refactor_notes.md]: i need to rewrite this section

File: llarp/CMakeLists.txt
Comment (majestrate) [reply llarp/CMakeLists.txt]: maybe, it is an optional system dep rather than a submodule so i didnt put it in external

Comment (majestrate) [reply llarp/CMakeLists.txt]: moved to external.

File: docs/refactor_notes.md
Comment (jagerman) [inline docs/refactor_notes.md:87]: I think src and dst need not be routers?

Comment (jagerman) [reply docs/refactor_notes.md:87]: i.e. maybe "src-peer-id" / "dst-peer-id" would be better terminology (unless I'm misunderstanding where talking to clients goes in this stack).

File: CMakeLists.txt
Comment (jagerman) [reply CMakeLists.txt]: I feel like this will come back to bite us in the ass, let's nuke the old name.

File: docs/refactor_notes.md
Comment (majestrate) [reply docs/refactor_notes.md:87]: src and dst are router id as that is what we care about here, we dont care about how we send it to them that is the wire layer's responsibility. really what it is is, "send this frame to router X" where X is addressed by their ed25519 public key.

Comment (majestrate) [reply docs/refactor_notes.md:87]: the link layer would then pick a wire protocol session that is established and then transfer it.

File: CMakeLists.txt
Comment (majestrate) [reply CMakeLists.txt]: yea, okay

**Comment by majestrate** [comment]
@jagerman looking to merge this today, needs approval.

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt:251]: I think we should split these off into a distinct `lokinet-rpc` library.

File: external/CMakeLists.txt
Comment (jagerman) [inline external/CMakeLists.txt:165]: This is currently enable/disabling a feature based on the presence of a dev package on your system.  I think we should feature-gate this behind a cmake option, and if enabled, require libcrypt.  We probably want that disabled by default for any of:
- static build
- Windows
- Android

Also probably worth leaving a "TODO" here to add it to StaticBuild.cmake (i.e. we should do it but defer that for now).

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt]: You can combine these PUBLIC/PRIVATE linking into the same command which makes it a bit easier to read:
```cmake
target_link_libraries(lokinet-consensus
    PUBLIC lokinet-util
    PRIVATE lokinet-context)
```

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt]: Likewise.

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt]: I think rather than making `lokinet-util` the "god" base target here (the the includes) we should instead of a `lokinet-common` (or `lokinet-base`?) INTERFACE library which carries the include path (and maybe other compiler flags) that every other lokinet-whatever need.

Comment (jagerman) [reply llarp/CMakeLists.txt]: Also I think with that change, most of the `PUBLIC` linking here (e.g. `lokinet-nodedb` -> `lokinet-cryptography`) can become `PRIVATE` instead.

**Review by jagerman — APPROVED** [review]
Left some comments; once those are addressed this looks good to merge.

**Comment by majestrate** [comment]
there was so many parts repeated that i ended up adding 2 helper functions to do the linking.
this will need a quick review again tomorrow.

(i will push this in the morning)

File: llarp/CMakeLists.txt
Comment (jagerman) [inline llarp/CMakeLists.txt]: CMake functions can have arguments which would save the ARGV manipulation here; any argument passed beyond the listed ones then go into ${ARGN}.  So this would simplify a bit to:
```suggestion
function(lokinet_link_lib libname)
  message(DEBUG "created target: ${libname}")
  enable_lto(${libname})
  target_link_libraries(${libname} PUBLIC lokinet-base ${ARGN})
  target_link_libraries(lokinet-amalgum INTERFACE ${libname})
endfunction()
```

**Comment by majestrate** [comment]
okay, ready for one final review

**Pull request merged**
